plugins/kubernetes/app/models/watchers/deploy_watcher.rb

```diff
@@ -1,119 +1,254 @@
+      start_watching
+    end
+
+    def self.watcher_symbol(project)
+      "deploy-watcher-#{project.id}".to_sym
+    end
+
+    def self.start_watcher(project)
+      watcher_name = watcher_symbol(project)
+      supervise as: watcher_name, args: [project]
+    end
+
+    def self.restart_watcher(project)
+      watcher = Actor[watcher_symbol(project)]
+      watcher.terminate if watcher and watcher.alive?
+      start_watcher(project)
+    end
+
+    def start_watching
+      sync_with_cluster
+      info "Start watching deployment for project: #{project.name}"
       async :watch
     end
 
     def watch
-      subscribe("#{@release.build.project_name}", :handle_update)
-    end
-
-    def handle_update(topic, data)
-      release_doc = release_doc_from_rc_name(topic)
-      release_updated = if data.object.kind == 'Event'
-                        handle_event_update(release_doc)
-                      else
-                        handle_pod_update(release_doc, data)
-                      end
-      if release_updated
-        send_event(role: release_doc.kubernetes_role.name,
-                   deploy_group: release_doc.deploy_group.name,
-                   target_replicas: release_doc.replica_target,
-                   live_replicas: release_doc.replicas_live,
-                   failed: release_doc.failed?)
-        end_deploy if deploy_finished?
+      subscribe("pod-events-project-#{@project.id}", :handle_event)
+    end
+
+    def handle_event(topic, message)
+      event = message.event
+      debug "Got message on topic: #{topic}. Type of message: #{event.kind}"
+
+      case
+      when event.kind == 'Event' then handle_cluster_event(message)
+      when event.kind == 'Pod' then handle_pod_event(message)
+      else error "Object kind not yet supported: #{notice.object.kind}"
       end
     end
 
     private
 
-    def handle_event_update(release_doc)
-      release_doc.failed? ? false : release_doc.fail!
+    # Gets the database in-sync with the Kubernetes Cluster
+    def sync_with_cluster
+      fetch_cluster_data
+      reconcile_old_releases
+      reconcile_db_with_cluster
     end
 
-    def handle_pod_update(release_doc, data)
-      pod_event = Events::PodEvent.new(data)
-      if pod_event.valid?
-        update_replica_count(release_doc, pod_event)
-        update_timeout(pod_event)
-        true
-      else
-        error 'invalid k8s pod event'
-        false
+    # From the target environment for the current Release, fetches all existing Pods from the corresponding
+    # Kubernetes Clusters and updates the internal data structures used by the watcher.
+    def fetch_cluster_data
+      env = release_env(last_release(project))
+      env.cluster_deploy_groups.each do |cdg|
+        cdg.cluster.client.get_pods(namespace: cdg.namespace, label_selector: "project_id=#{project.id}").each do |pod|
+          update_rc_pods(Kubernetes::Api::Pod.new(pod))
+        end
```

Comment:
Yeah... I was also thinking about that. And also that some of the code in this Watcher could, maybe, be refactored out to a separate class (e.g., KuberSynchronizationService).

I won't do it in this PR, but I'll keep a mental note to review this.